src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs

```diff
@@ -38,7 +38,7 @@ public virtual bool CanConvertFrom(ITypeDescriptorContext? context, Type sourceT
         /// Gets a value indicating whether this converter can convert an object to the given
         /// destination type using the context.
         /// </summary>
-        public virtual bool CanConvertTo(ITypeDescriptorContext? context, Type? destinationType)
+        public virtual bool CanConvertTo(ITypeDescriptorContext? context, Type destinationType)
```

Comment:
I looked into this more, and I think the advice I gave on https://github.com/dotnet/runtime/issues/63186#issuecomment-1004235914 wasn't 100% correct. Now that I look at it again, I think this should be annotated as:

```C#
public virtual bool CanConvertTo(ITypeDescriptorContext? context, [NotNullWhen(true)] Type? destinationType)
```

And then to fix the inconsistency, the above CanConvertTo overload should be annotated the same:

```C#
public bool CanConvertTo([NotNullWhen(true)] Type? destinationType)
```

Note that `ConvertTo` doesn't need to change. It doesn't allow for `null`, and will throw if someone passes `null`.

This way, people can write code like:

```C#
Type? someType = ...;

if (typeConverter.CanConvertTo(someType))
{
    var converted = typeConverter.ConvertTo(someType);
}
```

And they won't get any nullability warnings.

@krwq - since you originally added nullability here - do you have any thoughts?

Reply:
@eerhardt Thanks for the input. That does make sense. 

Should I expand the original ticket to also cover the `sourceType` parameter on CanConvertFrom?

Reply:
> Should I expand the original ticket to also cover the sourceType parameter on CanConvertFrom?

No. Let's open a new issue for CanConvertFrom. The scenario there is a bit different because presumably you've gotten the type from calling `something.GetType()`. ("From" implies you have something already that you are going to convert from.) I'd rather keep these separate, so we can get the CanConvertTo changes in without being held up by any issues with CanConvertFrom.

Reply:
@eerhardt Sure thing. Just let me know how you want to handle the ImageConverter-fix for CanConvertFrom.

Reply:
@eerhardt and @krwq Do you already have an update on the final decision on how to proceed with CanConvertTo (see https://github.com/dotnet/runtime/pull/63874#discussion_r786937186 , syntax update with `NotNullWhen` instead of not null)?

Reply:
I guess owner of this library would have better context on the intent of this method. If we don't want to throw on null destinationType then @eerhardt's suggestion (NotNullWhen) is correct and we avoid breaking changes. If we are ok with changing it to non-nullable type then that's fine too - it's a matter of deciding if we think it's likely a (consumer's) product bug to pass null or is it a valid scenario. I.e. if people sometimes pass in something like `obj?.GetType()` then I would go ahead with @eerhardt suggestion. IIRC when I originally annotated I changed this method multiple times as I wasn't sure and neither implementation nor tests nor docs convinced me either way.

Reply:
@MichaelKetting - I think we can go forward with my above suggestion. Any plans on moving this forward?

Reply:
@eerhardt Great, yes, will be updating the PR in the next couple of days / the weekend to go with the NotNullWhen solution.

Since the original null from way back changes did not update the docs, should I take the opportunity and do something there as well by adding parameter documentation and null guides?

Reply:
You can, but those docs are contained in a separate repo. Updating the xml doc above doesn't make it into the official docs.

Here are the official docs:
https://github.com/dotnet/dotnet-api-docs/blob/d90566928a7c5f6070f942a89d9be588647a7105/xml/System.ComponentModel/TypeConverter.xml#L349-L407

Reply:
@eerhardt Ah, okay, there they are :) This would then be a separate ticket and PR, correct, since they're separate repos? Which means I'll do the code first and once that's done and approved, I'll check out the docs.

Reply:
Yes, that would be a separate PR